src/box/iproto.cc

```diff
@@ -500,6 +500,19 @@ struct iproto_connection
 	 * connection.
 	 */
 	struct cmsg destroy_msg;
+	/**
+	 * Wait for shutdown readiness and close connection.
+	*/
+	struct ev_async after_shutdown;
+	/** Happened shutdown(socket, SHUT_RD) on input socket. */
+	bool socket_shutdown;
+	/**
+	* Connection support graceful shutdown. Is set only after start of
+	* shutdown process.
+	*/
+	bool graceful_shutdown;
+	/** Got shutdown packet from client. */
+	bool graceful_shutdown_got;
```

Comment:
But why do you even have these flags? We have connection state already. Can't you add new states to it? Was it too messy? I don't know if it will work, but it is certainly worth trying.